File: apps/mobile-app/utils/db/BaseRepository.ts
import * as dateFormatter from '@/utils/dateFormatter';

import NativeVaultManager from '@/specs/NativeVaultManager';

export type SqliteBindValue = string | number | null | Uint8Array;

/**
 * Interface for the core database operations needed by repositories.
 * Mobile app version is async as it communicates with native modules.
 */
export interface IDatabaseClient {
  executeQuery<T>(query: string, params?: SqliteBindValue[]): Promise<T[]>;
  executeUpdate(query: string, params?: SqliteBindValue[]): Promise<number>;
}

/**
 * Base repository class with common database operations.
 * Provides transaction handling, soft delete, and other shared functionality.
 *
 * Mobile-specific: All operations are async as they communicate with native modules.
 */
export abstract class BaseRepository {
  /**
   * Constructor for the BaseRepository class.
   * @param client - The database client to use for the repository
   */
  protected constructor(protected client: IDatabaseClient) {}

  /**
   * Execute a function within a transaction.
   * Automatically handles begin, commit, and rollback.
   * @param fn - The function to execute within the transaction
   * @returns The result of the function
   */
  protected async withTransaction<T>(fn: () => T | Promise<T>): Promise<T> {
    await NativeVaultManager.beginTransaction();
    try {
      const result = await fn();
      await NativeVaultManager.commitTransaction();
      return result;
    } catch (error) {
      await NativeVaultManager.rollbackTransaction();
      throw error;
    }
  }

  /**
   * Soft delete a record by setting IsDeleted = 1.
   * @param table - The table name
   * @param id - The record ID
   * @returns Number of rows affected
   */
  protected async softDelete(table: string, id: string): Promise<number> {
    const now = dateFormatter.now();
    return this.client.executeUpdate(
      `UPDATE ${table} SET IsDeleted = 1, UpdatedAt = ? WHERE Id = ?`,
      [now, id]
    );
  }

  /**
   * Soft delete records by a foreign key.
   * @param table - The table name
   * @param foreignKey - The foreign key column name
   * @param foreignKeyValue - The foreign key value
   * @returns Number of rows affected
   */
  protected async softDeleteByForeignKey(table: string, foreignKey: string, foreignKeyValue: string): Promise<number> {
    const now = dateFormatter.now();
    return this.client.executeUpdate(
      `UPDATE ${table} SET IsDeleted = 1, UpdatedAt = ? WHERE ${foreignKey} = ?`,
      [now, foreignKeyValue]
    );
  }

  /**
   * Hard delete a record permanently.
   * @param table - The table name
   * @param id - The record ID
   * @returns Number of rows affected
   */
  protected async hardDelete(table: string, id: string): Promise<number> {
    return this.client.executeUpdate(`DELETE FROM ${table} WHERE Id = ?`, [id]);
  }

  /**
   * Hard delete records by a foreign key.
   * @param table - The table name
   * @param foreignKey - The foreign key column name
   * @param foreignKeyValue - The foreign key value
   * @returns Number of rows affected
   */
  protected async hardDeleteByForeignKey(table: string, foreignKey: string, foreignKeyValue: string): Promise<number> {
    return this.client.executeUpdate(
      `DELETE FROM ${table} WHERE ${foreignKey} = ?`,
      [foreignKeyValue]
    );
  }

  /**
   * Check if a table exists in the database.
   * @param tableName - The name of the table to check
   * @returns True if the table exists
   */
  protected async tableExists(tableName: string): Promise<boolean> {
    const results = await this.client.executeQuery<{ name: string }>(
      `SELECT name FROM sqlite_master WHERE type='table' AND name=?`,
      [tableName]
    );
    return results.length > 0;
  }

  /**
   * Generate a new UUID in uppercase format.
   * @returns A new UUID string
   */
  protected generateId(): string {
    return crypto.randomUUID().toUpperCase();
  }

  /**
   * Get the current timestamp in the standard format.
   * @returns Current timestamp string
   */
  protected now(): string {
    return dateFormatter.now();
  }

  /**
   * Build a parameterized IN clause for SQL queries.
   * @param values - Array of values for the IN clause
   * @returns Object with placeholders string and values array
   */
  protected buildInClause(values: string[]): { placeholders: string; values: string[] } {
    return {
      placeholders: values.map(() => '?').join(','),
      values
    };
  }
}

File: apps/mobile-app/ios/VaultStoreKit/Repository/BaseRepository.swift
import Foundation
import VaultUtils

/// Base repository class with common database operations.
/// Provides transaction handling, soft delete, and other shared functionality.
public class BaseRepository {
    /// The database client used for executing queries.
    internal let client: DatabaseClient

    /// Initialize the repository with a database client.
    /// - Parameter client: The database client to use
    public init(client: DatabaseClient) {
        self.client = client
    }

    // MARK: - Transaction Helpers

    /// Execute a function within a transaction.
    /// Automatically handles begin, commit, and rollback.
    /// - Parameter operation: The function to execute within the transaction
    /// - Returns: The result of the function
    public func withTransaction<T>(_ operation: () throws -> T) throws -> T {
        try client.beginTransaction()
        do {
            let result = try operation()
            try client.commitTransaction()
            return result
        } catch {
            try? client.rollbackTransaction()
            throw error
        }
    }

    // MARK: - Soft Delete Helpers

    /// Soft delete a record by setting IsDeleted = 1.
    /// - Parameters:
    ///   - table: The table name
    ///   - id: The record ID
    /// - Returns: Number of rows affected
    @discardableResult
    public func softDelete(table: String, id: String) throws -> Int {
        let timestamp = DateHelpers.now()
        return try client.executeUpdate(
            "UPDATE \(table) SET IsDeleted = 1, UpdatedAt = ? WHERE Id = ?",
            params: [timestamp, id]
        )
    }

    /// Soft delete records by a foreign key.
    /// - Parameters:
    ///   - table: The table name
    ///   - foreignKey: The foreign key column name
    ///   - foreignKeyValue: The foreign key value
    /// - Returns: Number of rows affected
    @discardableResult
    public func softDeleteByForeignKey(table: String, foreignKey: String, foreignKeyValue: String) throws -> Int {
        let timestamp = DateHelpers.now()
        return try client.executeUpdate(
            "UPDATE \(table) SET IsDeleted = 1, UpdatedAt = ? WHERE \(foreignKey) = ?",
            params: [timestamp, foreignKeyValue]
        )
    }

    // MARK: - Hard Delete Helpers

    /// Hard delete a record permanently.
    /// - Parameters:
    ///   - table: The table name
    ///   - id: The record ID
    /// - Returns: Number of rows affected
    @discardableResult
    public func hardDelete(table: String, id: String) throws -> Int {
        return try client.executeUpdate(
            "DELETE FROM \(table) WHERE Id = ?",
            params: [id]
        )
    }

    /// Hard delete records by a foreign key.
    /// - Parameters:
    ///   - table: The table name
    ///   - foreignKey: The foreign key column name
    ///   - foreignKeyValue: The foreign key value
    /// - Returns: Number of rows affected
    @discardableResult
    public func hardDeleteByForeignKey(table: String, foreignKey: String, foreignKeyValue: String) throws -> Int {
        return try client.executeUpdate(
            "DELETE FROM \(table) WHERE \(foreignKey) = ?",
            params: [foreignKeyValue]
        )
    }

    // MARK: - Utility Methods

    /// Check if a table exists in the database.
    /// - Parameter tableName: The name of the table to check
    /// - Returns: True if the table exists
    public func tableExists(_ tableName: String) throws -> Bool {
        let results = try client.executeQuery(
            "SELECT name FROM sqlite_master WHERE type='table' AND name=?",
            params: [tableName]
        )
        return !results.isEmpty
    }

    /// Generate a new UUID in uppercase format.
    /// - Returns: A new UUID string
    public func generateId() -> String {
        return UUID().uuidString.uppercased()
    }

    /// Get the current timestamp in the standard format.
    /// - Returns: Current timestamp string
    public func now() -> String {
        return DateHelpers.now()
    }

    /// Build a parameterized IN clause for SQL queries.
    /// - Parameter values: Array of values for the IN clause
    /// - Returns: Tuple with placeholders string and values array
    public func buildInClause(_ values: [String]) -> (placeholders: String, values: [String]) {
        let placeholders = values.map { _ in "?" }.joined(separator: ",")
        return (placeholders, values)
    }
}
